diff --git a/node_modules/react-native-image-crop-picker/ios/src/ImageCropPicker.mm b/node_modules/react-native-image-crop-picker/ios/src/ImageCropPicker.mm
index 354ad09..9f88209 100644
--- a/node_modules/react-native-image-crop-picker/ios/src/ImageCropPicker.mm
+++ b/node_modules/react-native-image-crop-picker/ios/src/ImageCropPicker.mm
@@ -575,95 +575,100 @@ RCT_EXPORT_METHOD(openCropper:(NSDictionary *)options
                         });
                     }];
                 } else {
-                    [phAsset requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput * _Nullable contentEditingInput, NSDictionary * _Nonnull info) {
-                        [manager
-                         requestImageDataForAsset:phAsset
-                         options:options
-                         resultHandler:^(NSData *imageData, NSString *dataUTI, UIImageOrientation orientation, NSDictionary *info) {
-                            
-                            NSURL *sourceURL = contentEditingInput.fullSizeImageURL;
-                            
-                            dispatch_async(dispatch_get_main_queue(), ^{
-                                [lock lock];
-                                @autoreleasepool {
-                                    UIImage *imgT = [UIImage imageWithData:imageData];
-                                    
-                                    Boolean forceJpg = [[self.options valueForKey:@"forceJpg"] boolValue];
-                                    
-                                    NSNumber *compressQuality = [self.options valueForKey:@"compressImageQuality"];
-                                    Boolean isLossless = (compressQuality == nil || [compressQuality floatValue] >= 0.8);
-                                    
-                                    NSNumber *maxWidth = [self.options valueForKey:@"compressImageMaxWidth"];
-                                    Boolean useOriginalWidth = (maxWidth == nil || [maxWidth integerValue] >= imgT.size.width);
-                                    
-                                    NSNumber *maxHeight = [self.options valueForKey:@"compressImageMaxHeight"];
-                                    Boolean useOriginalHeight = (maxHeight == nil || [maxHeight integerValue] >= imgT.size.height);
-                                    
-                                    NSString *mimeType = [self determineMimeTypeFromImageData:imageData];
-                                    Boolean isKnownMimeType = [mimeType length] > 0;
-                                    
-                                    ImageResult *imageResult = [[ImageResult alloc] init];
-                                    if (isLossless && useOriginalWidth && useOriginalHeight && isKnownMimeType && !forceJpg) {
-                                        // Use original, unmodified image
-                                        imageResult.data = imageData;
-                                        imageResult.width = @(imgT.size.width);
-                                        imageResult.height = @(imgT.size.height);
-                                        imageResult.mime = mimeType;
-                                        imageResult.image = imgT;
-                                    } else {
-                                        imageResult = [self.compression compressImage:[imgT fixOrientation] withOptions:self.options];
-                                    }
+                    void (^handleImageData)(NSData*, NSString*, NSDictionary*, NSString*) = ^(NSData *imageData, NSString *dataUTI, NSDictionary *info, NSString *sourceURL) {
+                        dispatch_async(dispatch_get_main_queue(), ^{
+                            [lock lock];
+                            @autoreleasepool {
+                                UIImage *imgT = [UIImage imageWithData:imageData];
                                 
-                                    NSString *filePath = @"";
-                                    if([[self.options objectForKey:@"writeTempFile"] boolValue]) {
+                                Boolean forceJpg = [[self.options valueForKey:@"forceJpg"] boolValue];
                                 
-                                        filePath = [self persistFile:imageResult.data];
+                                NSNumber *compressQuality = [self.options valueForKey:@"compressImageQuality"];
+                                Boolean isLossless = (compressQuality == nil || [compressQuality floatValue] >= 0.8);
                                 
-                                        if (filePath == nil) {
-                                            [indicatorView stopAnimating];
-                                            [overlayView removeFromSuperview];
-                                            [imagePickerController dismissViewControllerAnimated:YES completion:[self waitAnimationEnd:^{
-                                                self.reject(ERROR_CANNOT_SAVE_IMAGE_KEY, ERROR_CANNOT_SAVE_IMAGE_MSG, nil);
-                                            }]];
-                                            return;
-                                        }
-                                    }
+                                NSNumber *maxWidth = [self.options valueForKey:@"compressImageMaxWidth"];
+                                Boolean useOriginalWidth = (maxWidth == nil || [maxWidth integerValue] >= imgT.size.width);
                                 
-                                    NSDictionary* exif = nil;
-                                    if([[self.options objectForKey:@"includeExif"] boolValue]) {
-                                        exif = [[CIImage imageWithData:imageData] properties];
-                                    }
+                                NSNumber *maxHeight = [self.options valueForKey:@"compressImageMaxHeight"];
+                                Boolean useOriginalHeight = (maxHeight == nil || [maxHeight integerValue] >= imgT.size.height);
+                                
+                                NSString *mimeType = [self determineMimeTypeFromImageData:imageData];
+                                Boolean isKnownMimeType = [mimeType length] > 0;
                                 
-                                    [selections replaceObjectAtIndex:index withObject:[self createAttachmentResponse:filePath
-                                                                                                            withExif: exif
-                                                                                                       withSourceURL:[sourceURL absoluteString]
-                                                                                                 withLocalIdentifier: phAsset.localIdentifier
-                                                                                                        withFilename: [phAsset valueForKey:@"filename"]
-                                                                                                           withWidth:imageResult.width
-                                                                                                          withHeight:imageResult.height
-                                                                                                            withMime:imageResult.mime
-                                                                                                            withSize:[NSNumber numberWithUnsignedInteger:imageResult.data.length]
-                                                                                                        withDuration: nil
-                                                                                                            withData:[[self.options objectForKey:@"includeBase64"] boolValue] ? [imageResult.data base64EncodedStringWithOptions:0]: nil
-                                                                                                            withRect:CGRectNull
-                                                                                                    withCreationDate:phAsset.creationDate
-                                                                                                withModificationDate:phAsset.modificationDate
-                                                                                      ]];
+                                ImageResult *imageResult = [[ImageResult alloc] init];
+                                if (isLossless && useOriginalWidth && useOriginalHeight && isKnownMimeType && !forceJpg) {
+                                    // Use original, unmodified image
+                                    imageResult.data = imageData;
+                                    imageResult.width = @(imgT.size.width);
+                                    imageResult.height = @(imgT.size.height);
+                                    imageResult.mime = mimeType;
+                                    imageResult.image = imgT;
+                                } else {
+                                    imageResult = [self.compression compressImage:[imgT fixOrientation] withOptions:self.options];
                                 }
-                                processed++;
-                                [lock unlock];
                                 
-                                if (processed == [assets count]) {
+                                NSString *filePath = @"";
+                                if([[self.options objectForKey:@"writeTempFile"] boolValue]) {
+                                    
+                                    filePath = [self persistFile:imageResult.data];
+                                    
+                                    if (filePath == nil) {
+                                        [indicatorView stopAnimating];
+                                        [overlayView removeFromSuperview];
+                                        [imagePickerController dismissViewControllerAnimated:YES completion:[self waitAnimationEnd:^{
+                                            self.reject(ERROR_CANNOT_SAVE_IMAGE_KEY, ERROR_CANNOT_SAVE_IMAGE_MSG, nil);
+                                        }]];
+                                        return;
+                                    }
+                                }
                                 
-                                    [indicatorView stopAnimating];
-                                    [overlayView removeFromSuperview];
-                                    [imagePickerController dismissViewControllerAnimated:YES completion:[self waitAnimationEnd:^{
-                                        self.resolve(selections);
-                                    }]];
-                                    return;
+                                NSDictionary* exif = nil;
+                                if([[self.options objectForKey:@"includeExif"] boolValue]) {
+                                    exif = [[CIImage imageWithData:imageData] properties];
                                 }
-                            });
-                        }];
+                                
+                                [selections replaceObjectAtIndex:index withObject:[
+                                    self createAttachmentResponse:filePath
+                                    withExif: exif
+                                    withSourceURL:sourceURL
+                                    withLocalIdentifier: phAsset.localIdentifier
+                                    withFilename: [phAsset valueForKey:@"filename"]
+                                    withWidth:imageResult.width
+                                    withHeight:imageResult.height
+                                    withMime:imageResult.mime
+                                    withSize:[NSNumber numberWithUnsignedInteger:imageResult.data.length]
+                                    withDuration: nil
+                                    withData:[[self.options objectForKey:@"includeBase64"] boolValue] ? [imageResult.data base64EncodedStringWithOptions:0]: nil
+                                    withRect:CGRectNull
+                                    withCreationDate:phAsset.creationDate
+                                    withModificationDate:phAsset.modificationDate
+                                ]];
+                            }
+                            processed++;
+                            [lock unlock];
+                            
+                            if (processed == [assets count]) {
+                                
+                                [indicatorView stopAnimating];
+                                [overlayView removeFromSuperview];
+                                [imagePickerController dismissViewControllerAnimated:YES completion:[self waitAnimationEnd:^{
+                                    self.resolve(selections);
+                                }]];
+                                return;
+                            }
+                        });
+                    };
+                    
+                    [manager requestImageDataForAsset:phAsset options:options resultHandler:^(NSData *imageData, NSString *dataUTI, UIImageOrientation orientation, NSDictionary *info) {
+                        BOOL isInCloud = [info[PHImageResultIsInCloudKey] boolValue];
+                        BOOL isCropping = [[self.options objectForKey:@"cropping"] boolValue];
+                        if(isInCloud || isCropping) {
+                            [phAsset requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput * _Nullable contentEditingInput, NSDictionary * _Nonnull info) {
+                                handleImageData(imageData, dataUTI, info, [contentEditingInput.fullSizeImageURL absoluteString]);
+                            }];
+                        } else {
+                            handleImageData(imageData, dataUTI, info, nil);
+                        }
                     }];
                 }
             }
@@ -687,14 +692,7 @@ RCT_EXPORT_METHOD(openCropper:(NSDictionary *)options
                     });
                 }];
             } else {
-                [phAsset requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput * _Nullable contentEditingInput, NSDictionary * _Nonnull info) {
-                    [manager
-                     requestImageDataForAsset:phAsset
-                     options:options
-                     resultHandler:^(NSData *imageData, NSString *dataUTI,
-                                     UIImageOrientation orientation,
-                                     NSDictionary *info) {
-                        NSURL *sourceURL = contentEditingInput.fullSizeImageURL;
+                void (^handleImageData)(NSData*, NSString*, NSDictionary*, NSString*) = ^(NSData *imageData, NSString *dataUTI, NSDictionary *info, NSString *sourceURL) {
                         NSDictionary* exif;
                         if([[self.options objectForKey:@"includeExif"] boolValue]) {
                             exif = [[CIImage imageWithData:imageData] properties];
@@ -707,13 +705,24 @@ RCT_EXPORT_METHOD(openCropper:(NSDictionary *)options
                             [self processSingleImagePick:[UIImage imageWithData:imageData]
                                                 withExif: exif
                                       withViewController:imagePickerController
-                                           withSourceURL:[sourceURL absoluteString]
+                                           withSourceURL:sourceURL
                                      withLocalIdentifier:phAsset.localIdentifier
                                             withFilename:[phAsset valueForKey:@"filename"]
                                         withCreationDate:phAsset.creationDate
                                     withModificationDate:phAsset.modificationDate];
                         });
-                    }];
+                };
+                
+                [manager requestImageDataForAsset:phAsset options:options resultHandler:^(NSData *imageData, NSString *dataUTI, UIImageOrientation orientation, NSDictionary *info) {
+                    BOOL isInCloud = [info[PHImageResultIsInCloudKey] boolValue];
+                    BOOL isCropping = [[self.options objectForKey:@"cropping"] boolValue];
+                    if(isInCloud || isCropping) {
+                        [phAsset requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput * _Nullable contentEditingInput, NSDictionary * _Nonnull info) {
+                            handleImageData(imageData, dataUTI, info, [contentEditingInput.fullSizeImageURL absoluteString]);
+                        }];
+                    } else {
+                        handleImageData(imageData, dataUTI, info, nil);
+                    }
                 }];
             }
         }];
